### ISSUE TITLE
bisync: performance

### ISSUE THREAD

**Issue by crocinsocks** [rclone/rclone#7332]
### setup

11,265 files - 290.743MiB (90MB in ~5 files the rest are smaller and more equal in size)

Results the same regardless of Wasabi or iDrive (iDrive typically performs a little faster in every test but only ~5-10% on average)

Remote is using crypt and base32768 encoding. 

BW: ~1.5MiB UP ~5MiB DOWN

### default/common flags

After testing best performing filters are: --transfers 64 --checkers 128 --fast-list

Along with this (--filter-from) or (--filters-file) is also used

e.g. --transfers 64 --checkers 128 --fast-list --filter-from/--filters-file ..

### check

~8s on average and between 6.8s to 9.6s

### ls

(using `time` as no info given with -v)

ls: 0m4.088s
lsl: 7m36.975s

### sync

upload:
	initial: 3m57.0s
	subsequent: 18.6s
	
download:
	initial: 2m9.0s
	subsequent: 17.8s

### copy

upload:
	initial: 4m0.0s
	subsequent 18.6s
	
download:
	initial: 2m6.5s
	subsequent: 16.8s

### bisync

These are some of the best and would say are usually a little longer by upto 5 mins this is likely when resync is run when data is already present vs copying the data itself.. There are no changes made to files when moving between a --resync and normal run.  

defaults:
	resync: 10m47.0s
	normal: 6m44.7s

defaults with ignore checksum: (--ignore-listing-checksum --ignore-checksum)
	resync: 10m36.5s
	normal: 6m31.9s

### overview

So as you can see bisync is orders of magnitude slower than `copy` or `sync` . `ls` is incredibly fast as is `sync` and `copy` in both directions (e.g. uses up almost 100% of bandwidth)

I only ran the test of `lsl` and `ls` at the very end and now thinking that bisync relies on lsl as this so incredibly slow (as I believe bisync does not use checksum so would need `lsl` info to compare?).

I can do a local to crypt (local) bisync in 15s (--resync) and 8.7s for subsequent. 

Is there something I am doing wrong? If not, are there plans/options available to rclone to improve the speed of bisync so that it is more in line with 2 x sync/copy commands (e.g. one upload, one download) or is this a limitation that is unlikely to be overcome? Also, if this is due to lsl taking so long is there a way to speed up lsl? 

Thanks to you all for your hard work and a special thank you @nielash for all you've been doing on bisync and continue to do. I as many others massively appreciate it and when this tool is out of experimental stages it will be a game changer for so many. If I had any free time or money right now I would help out but unfortunately the best I can do is give you some benchmarks and user feedback.

Thanks

**Comment by nielash** [comment]
Thank you for the benchmarks and feedback! You are right -- `bisync` is currently much slower than `copy` or `sync`. This is for a few reasons, most of which I think are solvable. My multi-step plan to address this includes:

- [x] Apply filters correctly during deletes (fixed in `v1.64.0` https://github.com/rclone/rclone/pull/6971/commits/26ff813ff3bcf1755631f840e6b4464083a8d80d )
- [ ] Build and compare initial listings concurrently using March 
This should help a lot. Currently, bisync builds a full listing for Path1, then a full listing for Path2, then compares them -- and each of those tasks must finish before the next one can start. That's silly, as we could easily be doing most of these tasks concurrently, using the same March infrastructure that `check` and `sync` use. This would also solve another important issue I talked about [here](https://forum.rclone.org/t/bisync-bugs-and-feature-requests/37636#:~:text=4.%20Listings%20should%20alternate%20between%20paths%20to%20minimize%20errors).
- [ ] Merge copies and deletes
Currently bisync makes its changes in a 4-step process:

1. `copy` 2 to 1
2. `copy` 1 to 2
3.  `delete` 2 to 1
4.  `delete` 1 to 2

I believe we can condense this to just 2 steps:

1. `sync` 2 to 1
2. `sync` 1 to 2

Both syncs would be filtered for only the individual files that need to be copied or deleted in that direction. I already have a local proof-of-concept for this that works beautifully, so I think it should be coming relatively soon. Incidentally, this change also paves the way for [allowing bisync to support `--track-renames`](https://github.com/rclone/rclone/issues/5690) (which is supported only for `sync`, not `copy`.)

- [ ] Eliminate the second listing
This is probably the one that will make the most difference in terms of speed. Currently, if there are changes to sync, bisync lists each directory twice: once before the sync and once after. The second one causes [quite a lot of problems](https://github.com/rclone/rclone/issues/5676), in addition to making each run much slower. I am currently working on a way to eliminate it by getting the underlying sync operation to [report back a list of what it changed](https://github.com/rclone/rclone/issues/7282). This project is pretty far along, and the results are very encouraging so far! 

When all of the above are done, `bisync` will probably still be slightly slower than `sync` (because bisync has to do an extra listing, and 2 syncs instead of 1), but the difference should be much less dramatic than it is today. 

For `bisync` to ever truly be as fast as `sync`, it would have to (at minimum) merge the two sync operations together (or at least do them concurrently), and also merge the initial listing into the sync operation (similar to how checkers and transfers currently work during a `sync` -- note how the transfers are able to start working before all the checkers have completed!) These particular optimizations are not currently planned, but I do think they might be possible someday. 

In the meantime, one thing you might consider to speed up your runs slightly is `--check-sync=false`. `--check-sync` has some [issues](https://github.com/rclone/rclone/issues/7270) anyway at the moment, so in my opinion, you might be better off disabling it and replacing it with an occasional [check](https://rclone.org/bisync/#concurrent-modifications), until this is more properly fixed in a future release. Disabling it will make the run slightly faster, because it will skip the step of re-loading both listings into memory at the end and comparing them.

Thank you also for the kind words of encouragement! 😃

**Comment by crocinsocks** [comment]
This is awesome and sounds very promising! Thank you for the detailed response it was very useful as I'm sure for others stumbling across this.

In particular, implementing March and the secondary listing to bisync sounds very promising and can see the vast improvement of comparing the files in real-time over a delayed listing and further prevent this locking/corruption issue when there are changes made during sync. Also, as I understand this is one of the major issues left to bring bisync out of experimental?

Can't wait to have this working! For now, I will also try with `--check-sync=false` as ultimately I plan on running this at regular intervals anyways.

Would these changes also help with the following situation (perhaps this is not the best place to ask this or maybe a new issue or forum?):

I have a systemd service that runs bisync on a timer/repeat. If I shutdown my computer and bisync is running it would require a full resync as pressing Ctrl+C would do when running in terminal (as far as I can tell - or at least seems to be this way from my tests?). I haven't had any luck using `--resilient` in this case either. This means for most of the longer running bisync (and occasionally shorter syncs) will fail requiring a --resync every reboot (or at least I'd have to assume this). This is also difficult as you couldn't automate this as I would need to know whether to resync as push or pull of the remote. Would be fantastic if bisync can recover from a failed or rather cancelled sync without requiring a resync. Would any of these upcoming changes fix this too? Or maybe my setup is wrong? 

Good luck with it all and thank you again 🙏

**Comment by nielash** [comment]
> I have a systemd service that runs bisync on a timer/repeat. If I shutdown my computer and bisync is running it would require a full resync as pressing Ctrl+C would do when running in terminal

An interrupted run is tough for `bisync` to recover from cleanly, because of its bi-directional nature. When `sync` is interrupted, it can be safely retried because the source is still intact, even if the dest was left in a messy state. But with `bisync`, _both_ sides could be left in a messy state, if interrupted. Before `v1.63.0`, I would have said clean recovery is impossible, due to the possibility (on certain backends) of leaving a partially-written file on one path, which would then get erroneously copied to the other path on the next run. But now that we have ["atomic downloads"](https://github.com/rclone/rclone/pull/6913), I think this is worth another look. 

One challenge would be bisync's logic surrounding [lock files](https://rclone.org/bisync/#lock-file) and the saving of the "new" state for use by the next run. Currently bisync finalizes these only after a successful sync operation. Doing this after every individual transfer probably wouldn't be practical -- it would add too much overhead and too many writes to disk. The cleaner solution is probably the "retry from last successful snapshot on error" idea that I mentioned [here](https://github.com/rclone/rclone/issues/5692#issuecomment-1689973729). Essentially what this would mean is: if the most recent listing is untrusted (because, in this case, it was interrupted), and `--resilient` is specified, we would look instead to the one before that, and diff the current state against the last "trusted" state. This would effectively make it possible to recover from these kinds of errors by "retrying" on the next run.

We would also need some way of distinguishing between lock files left by aborted runs vs. ones that are still running. Perhaps lock files should be considered "expired" after X number of minutes (configurable with a setting)? (And, thinking ahead: perhaps they should also get "renewed" every X-1 minutes by runs that are still running.)

In the meantime, my recommendation is: try to wait to shut down your computer until you know the bisync run has completed. (FWIW: I have mine scripted to send me a push notification when it starts, and another with the exit code when it finishes.)

**Comment by crocinsocks** [comment]
Great news! I've taken a look at how lock files work and in my opinion the best option would be to check the PID alongside a regex similar to `*.rclone.*.PATH1.*.PATH2` in the case the PID is reused for another application (rclone included). This would save any 'hackery' solution and get straight to the source. 

With regards to Sessions, I believe this is the best way and I do not see a reason you couldn't have a partial session where every 30s-1m the session data is updated so in a recovery situation rather than going back to the last complete session you can continue from the last complete session + partial session. However, this might be overkill.

In regards to handling this issue.. I have many bisync's running so it would be too hard to handle such a situation manually or with push notification as an aid.. So my current approach is going to have to be to catch any shutdown requests. Then, I would regularly (every 5s or so) watch the systemd process to be in the off state, stop the process so it does not run again and repeat for each rclone service. Only when all of them have been stopped then I would continue shutdown/restart/sleep etc... Perhaps, a push notification to let me know if one requires a re-sync (power cut for example) so there is no ambiguity.

**Comment by nielash** [comment]
> in my opinion the best option would be to check the PID alongside a regex similar to `*.rclone.*.PATH1.*.PATH2` in the case the PID is reused for another application (rclone included). This would save any 'hackery' solution and get straight to the source.

This would work well if you're running bisync on only one machine. However, one of our long-term goals is to make bisync more portable by allowing any remote path to be used as the `--workdir` (currently it is limited to local paths). One of the benefits of storing lock files on a non-local remote is that you could then easily [use bisync to keep several devices in sync with a cloud storage provider](https://github.com/rclone/rclone/issues/5689#issuecomment-1685612894), without the risk of [syncing the same path simultaneously from different locations](https://rclone.org/faq/#using-rclone-from-multiple-locations-at-the-same-time). The PID is less useful in this scenario, because each machine would not know the PIDs from the other machines. (They also would not be unique, but that could possibly be solved by adding some kind of machine/location identifier.)

> I do not see a reason you couldn't have a partial session where every 30s-1m the session data is updated so in a recovery situation rather than going back to the last complete session you can continue from the last complete session + partial session. However, this might be overkill.

This is an interesting idea -- sort of analogous to an "autosave" feature. 30s-1m could be reasonable with a local `--workdir`, but may not be for other backends -- particularly ones that charge per API call. It would also slow down the run. But perhaps the option should exist as a setting, and then each user can decide what's right for them. I'm not sure that this method would have much advantage over the "retry from last successful" method... the one scenario I can think of where it would make a difference is:

1. File changes on one side
2. File is synced successfully during a run that later aborts
3. Same file changes AGAIN before the next run
4. File is synced again on next run

Absent step 3, the results should be identical with either method. (With the "retry..." method, bisync should say "I see that the file has changed on both sides since the last successful run, but since the file currently matches on both sides, I can skip this and move on.")

> I have many bisync's running so it would be too hard to handle manually or with push notification as an aid.. So my current approach is going to have to be to catch any shutdown requests...

I wonder if what you ultimately want here is something more like #966 (a way to tell all your running bisync processes to gracefully wrap-up early, when you want to shut down.) (See also: #1522)

**Comment by crocinsocks** [comment]
I only found rclone 2 weeks ago and with the general complexity of this topic do forgive my ignorance that my come from the below.

I can only see the need for --workdir to be on a remote when syncing the *same* remotes from a stateless application e.g. a cronjob on k8s instance. Otherwise, a simple lock file would suffice and yes setting an expiry time and the running process updating the expiry would work. Syncing the rest of workdir seems overkill (with no other advantages?) whilst also running the risk of duplicate IDs, more requests where not needed and I'm sure other problems too with regards to multiple PCs using the same workdir. When the lockfile is local I believe it should function using PID and regex as there seems no harm in a simple if statement to handle these two cases.

Though saying all this, I personally would never want my setup to work this way. Blocking another sync in a multi pc setup seems antiquated and a longer running syncs would damage a functional workflow. As far as I can tell the case you are you are describing here is conflict resolution in the very rare chance that the same path is uploaded at precisely the same time. In my case as I do not have request costs I would prefer a lockfile for each upload or better yet perhaps a simple check before committing an upload that the file has not changed (if not handle conflict) or even the use of a .part file / tmp upload dir (though I know there are some issues with move on certain remotes). Regardless, I think there are many ways to handle this rare situation that do not involve blocking a PC to one sync at a time nor require a remote workdir.

If using lock files a simple flag `--lock-remote 'global' | 'file'` ` --global-lock-name '.rclone-lock' (default)`

After, looking into how rclone handles conflict resolution, I would say a small change is needed in that the newest file should not be renamed rather kept as the master and any conflicts are made using .conflict1, .conflict2 and so on over _PATH1, _PATH2 so it is more inline with the typical two-way sync. Further this would not not limit multiple conflicts and help with the potential issue of file name length when using _PATH1 & _PATH2. 

Am I correct in thinking that the only [race conditions](https://rclone.org/faq/#using-rclone-from-multiple-locations-at-the-same-time) in bisync is the example above which is the same file being written at same time from two machines where remote file has not changed and yet both files are different on each PC (I understand others exist in sync as it is stateless)? Ultimately, I have two machines running bisync at the same time (give or take) with the same sync command e.g. `rclone bisync ~/Documents remote@documents:` and fail to see how this would cause any problems especially once the listing/operation process is changed "Listings should alternate between paths to minimize errors". At this stage, I have an archive so am okay with the odd corruption here and there however it would seem unlikely that I would ever see such a thing? Especially, as I typically would never change the same file on both machines prior to a sync.

Also, I should probably comment on #5681 directly however just want to add that the way I would have expected rclone to run the initial sync is with: `rclone copy --update Path1 Path2 (--create-empty-src-dirs) && rclone copy --update Path2 Path1 (--create-empty-src-dirs)`. For example, this what would happen if I was using Dropbox, OneDrive, GDrive, etc e.g. the newest file will always survive on init and where the local directory already has data. Currently, it seems Path1 will always take precedent which is the last thing I believe anyone would want from a bisync. Further, I believe this should be the default behaviour and a legacy flag a made for those that would need it (as I struggle to understand who would want this and if so perhaps a manual `sync` is better for these people and then you could remove the old functionality all together?). Ultimately, the end goal for bisync should be a --resync is never required after init and failures are handled using conflict resolution or otherwise. I know this is probably easier said than done but a lot of changes you are already working on should get this extremely close. 

With regards to session, my view is if the workdir is local then this should happen regularly at reasonable and predefined interval and perhaps disabled by default when remote.  In this case I'd probably even lower default for local to 10-20s. e.g. `--session-interval 0 (disabled & default with remote --workdir) | 15s (default with local --workdir)`

The issue with something like a graceful shutdown for rclone is most applications are only given a moments notice between SIGTERM and a SIGKILL which is unlikely enough time to let any current transactions complete though enough to likely create a safe recovery point. Also, I would prefer my computers to be fully up-to-date on shutdown so when switching between machines I know I will always have the latest copy. I've implemented the method described in my previous comment which works well for my use case. If/once bisync can handle a SIGTERM & SIGKILL event and recover without a resync this would likely work well for most use cases. And further with the possibility of dramatic speed increase most regularly run bisync will likely be up-to-date on shutdown unless the user has made very large changes at which point they'd know to leave the machine on.

**Comment by nielash** [comment]
Thank you so much for this detailed feedback! It is really helping me gain insight into bisync use cases that are different from my own, which is quite valuable 🙂 (As an aside, I wonder if it might be helpful to start a "bisync user feedback" group in the forum or something -- it would be really helpful to be able to easily survey a group from time to time on these kinds of design questions.)

A few responses to your comments:
> What are the benifits of a remote --workdir? The only example I can think of is to make syncing the same two remotes in stateless application easier? If the only requirement is to lock a sync it seems like a lot of work with greater potential for error if multiple PCs with different listings use the same remote workdir.

In a word: portability. This question sort of implies that at least one of the bisync paths is a `local` remote, and while I agree that's probably the most common use case for bisync, it's not technically a requirement. For example, bisync could just as easily be used to sync a `drive` remote with a `dropbox` remote. In this kind of scenario, it makes a little less sense to require that bisync must always be run from the same machine. Perhaps I'm traveling with my laptop and away from my main desktop machine -- if the listing files are in the cloud, I can seamlessly pick up on one machine where I left off on another. It also just strikes me as more "rclonic" -- in general, rclone is designed to be as stateless as possible, and avoid relying on sidecars, resource forks, or local databases. Bisync is currently one of the only rclone commands that does so (out of necessity, as bidirectional sync can never be completely stateless). So, in my (subjective) opinion, allowing a non-local `--workdir` brings us a little bit closer to rclone's platform-agnostic core principles, and the (unreachable) goal of having full feature parity between backends.

It should also be noted that adding support for non-local `--workdir` does not mean removing any existing support for `local` -- so really nothing would change for users that wish to continue using a local `--workdir`. They can ignore this new feature and carry on like they did before.

> After, looking into how rclone handles conflict resolution, I would say a small change is needed in that the newest file should not be renamed rather kept as the master

One of the [most common complaints](https://github.com/rclone/rclone/issues/5679#:~:text=The%20fact%20that,one%20side%20only.) about bisync is that it currently [refuses to operate on backends that lack modification time support](https://rclone.org/bisync/#modification-time). This is going to be solved soon by adding support for [`--checksum`](https://github.com/rclone/rclone/issues/5683) and [`--size-only`](https://github.com/rclone/rclone/issues/5684). Once this is the case, there will no longer really be a concept of "newest" in a file-conflict situation -- all we can say is whether a file changed on both paths (since the last run), and whether or not the current version on one path is [equal](https://github.com/rclone/rclone/pull/6971/commits/add917cfac363bd49ac718123f29f764ea5531c0) to the version on the opposite path. So, this would present a challenge to what you are proposing -- sometimes there is no "newest", only "different". I could imagine possibly implementing what you are suggesting as an optional flag for the rest of the time (when we do have modtime available on both sides), although personally I'm not sure I would use it. To me, the `..path1` / `..path2` rename is a helpful alert that something went wrong and needs my attention to resolve, so silently skipping this for one of the paths would give me a false sense of security. That said, I have no objection to implementing features I don't personally plan to use (I much prefer giving users the choice whenever possible), so if a sizable number of users want this, I'm happy to add the option. (This is a good example of something I'd love to poll the group about!)

Also worth noting: even when both paths do support modtime, "deletes" still never have a modtime, just by their nature. (All we can say is that a file used to be there and now is not -- we can't say _what time_ it was deleted, or whether it was deleted before or after the other path's version was last modified.)

> any conflicts are made using .conflict1, .conflict2 and so on over _PATH1, _PATH2 so it is more inline with the typical two-way sync. Naming the files ..path1 & ..path2 limits the number of conflict and can lead to further issues when multiple PCs are used at the same time. Should I open another issue for this?

I'm not sure if I understand what you mean about this one. How does `..path1` / `..path2` limit the number of conflicts, and how would `.conflict1` / `.conflict2` solve that? Does this question rely on the the prior suggestion (of renaming only one side) being implemented? To make sure we're on the same page about the _current_ behavior: the first conflict for file `x` results in `x..path1` (for the Path1 side) and `x..path2` (for the Path2 side). If the Path1 conflict were to have yet another conflict after that (and I'm not sure why that would happen, as it would mean opening `x..path1` and editing it on both sides without renaming it first -- but let's just assume it did), you would then get `x..path1..path1` and `x..path1..path2`, in addition to the pre-existing `x..path2` which would remain unchanged. You could repeat this any number of times (limited only by filename length limitations from your backend...eventually.) I'm guessing that maybe what you would prefer is for subsequent conflicts to be named `x..conflict3`, `x..conflict4`, `x..conflict5`, etc. but my question then would be: how would you tell whether `x..conflict5` came from Path1 or Path2? And how do you determine the winner when neither side can be considered "newest"?

> As bisync is stateful rather than stateless the only race condition I could think of is the same path being written at the same time?

By "path" do you mean one individual file? That is certainly a concern, but I think it's also important to consider the risk at the _directory_ level. When rclone walks directories recursively, the [order can be unpredictable](https://rclone.org/docs/#limitations) due to concurrency, among other things -- there's no guarantee that `bisync A B` and `bisync A C` will both list the directories of `A` in the same order. (The tradeoff is that concurrency is what makes rclone so fast.) So given that, now imagine a use case where the sync needs to be atomic at not just the _file_ level, but also the _directory_ level. For example, something like a source code repository, where _all_ files in the tree must be from version 1 or else _all_ files in the tree must be from version 2. You can't end up with one subdirectory from version 1 and another from version 2 in the same repo -- the application won't work. This is the kind of problem we could miss if we only worry about conflicts between individual files and consider each file an island. So, while you're right that the chance of writing any individual file at the exact same time from two locations is fairly small, the chance of doing the same thing to a _directory_ is much larger -- the race condition window could be even hours or days-long, if the files are particularly large. A directory-level conflict could occur without any of its containing files having a file-level conflict. Most concurrent applications solve this problem with a mutex or other kind of locking mechanism, and that is why bisync uses "lock files", although the implementation is imperfect at the moment because the mutual exclusion enforcement doesn't extend to other machines (nor does it detect tree overlaps even within the same machine.) To give credit where credit's due, though, the original designer of bisync (not me, to be clear) realized this problem and made sure to include this warning in [the docs](https://rclone.org/bisync/#lock-file):
>>"**Note** that while concurrent bisync runs are allowed, be _very cautious_ that there is no overlap in the trees being synched between concurrent runs, lest there be replicated files, deleted files and general mayhem."

Another common scenario where mutual exclusion matters at the directory-level is a "project folder" like the kind often created by media editing applications (Pro Tools, Adobe Premiere / Illustrator, Apple Logic Pro, etc.) These apps expect all the source media files to be from the same project version (for example, video editors often throw errors if the video file they find at a particular path doesn't match the length of the clip in the timeline.) This kind of data corruption is not something we can catch by looking solely for conflicts between individual files.

> With regards to session, my view is if the workdir is local then this should happen regularly at reasonable and predefined interval and perhaps disabled by default when remote

I think we may be using the word "sessions" to mean different things (I'm using it as defined [here](https://github.com/rclone/rclone/issues/5678)), but setting that aside, I think we generally agree on the substance -- it would be ideal for bisync to be more prepared for the possibility of an unexpected interruption, and it should require `--resync` as little as possible. I think where we might differ slightly is our preference when faced with the inevitable tradeoff between convenience and safety. While I'm no fan of being asked to `--resync`, I still would rather be asked than have bisync take a "best guess" on my behalf and guess wrong. One of the main reasons I chose bisync (and rclone as a whole) in the first place is because of how it prioritizes safety where other sync tools prioritize convenience or resilience. I feel comfortable entrusting important files to bisync precisely because it will err on the side of stopping and asking me for help when its only alternative is to guess. I fully acknowledge that some users will prefer a different tradeoff and may be fine settling for a standard where data corruption is just "unlikely" -- but I can't assume that every user will be ok with that. So, I plan to maintain bisync's original philosophy of being "(too) safe by default", but changeable with flags for advanced users. (My recent addition of [`--resilient` mode](https://rclone.org/bisync/#resilient) is an example.)

To sum up where I think I am regarding crash recovery: I think there are definitely some meaningful improvements we can and should make, and in particular I like your idea of an "autosave"-like feature that is disabled by default when `--workdir` is remote (and I'd say possibly when local too -- unless careful testing finds the performance overhead to be basically negligible.) To be totally transparent, I think it will likely take me awhile to get to it, as I think bisync has several other more pressing issues, and I also don't find "don't pull the plug while it's running" to be an unreasonable or excessively burdensome expectation in the meantime (but again, that's not to say we shouldn't try to improve it -- we should.) You are also of course welcome to submit a Pull Request at any time, if you would like to. You might also consider checking out some [bisync alternatives](https://rclone.org/bisync/#references) (such as [syncrclone](https://github.com/Jwink3101/syncrclone)) that make some different choices with regard to some of these tradeoffs.

> I've implemented the method described in my previous comment which seemingly works well for my use case.

Glad to hear that! I'd also be interested to know a little more about your use case (only if you're inclined to share) and why it requires so many different bisync processes to be running simultaneously. For example -- could it not be solved by [combining](https://rclone.org/combine/) the various paths into one tree and then bisyncing them all in one bisync run? Also, how frequently do you schedule the runs, and what's the longest they normally take? (I think this info may help me conceptualize why curing the interruptions would be more practical than preventing them.)

Thank you again for the thoughtful feedback! You've challenged my thinking in a good way 🙂

**Comment by crocinsocks** [comment]
## Workdir

This is exactly the use case I was suggesting here, however I should have said stateless manner over application.

> The only example I can think of is to make syncing the same two remotes in stateless application easier? 

I was thinking something like k8s cronjob however your example of laptop to desktop and keeping the same two remotes (e.g. remoteA: remoteB:) is just as valid. This could already be solved by using a systemd service for example and using StartPre to update local workdir from a remote, Start to run bisync, StartPost to update local workdir to a remote. Though sure adding this functionality could make life easier provided one does not do what I talk about later in this section. 

However, I should have been more clear with my question as I was only asking due to your comment:  

> This would work well if you're running bisync on only one machine. However, one of our long-term goals is to make bisync more portable by allowing any remote path to be used as the --workdir (currently it is limited to local paths). One of the benefits of storing lock files on a non-local remote is that you could then easily use bisync to keep several devices in sync with a cloud storage provider, without the risk of syncing the same path simultaneously from different locations. The PID is less useful in this scenario, because each machine would not know the PIDs from the other machines. (They also would not be unique, but that could possibly be solved by adding some kind of machine/location identifier.)

The comment here made me go well hold on, why would you store the lock file in the --workdir and how would having a remote --workdir help keep multiple pc in sync. I can only see it making it worse. If I have two computers syncing `~/Documents remote:` having a remote --workdir would require a complete change to --workdir to handle unique ids and runs the risk of duplicate IDs, more requests where not needed and I'm sure other problems too. Also, a lock file is to do with the remote and not workdir. Of course, I understand you would never remove a local --workdir. When syncing multiple PCs it should be a limitation that the workdir must be local as there can be no benefit to storing it remotely and if one really wants to (for some unknown reason at least that I can think of?) it must be a different remote path ergo --workdir of any other pc.

## Mod time & Conflicts

By very nature and for bisync to be fully functional requires a backend to support mod time. That's not to say that workarounds can't be put in place but simply when using bisync one should always be advised to use it with a backend that supports modtime. Checksum should really on be used to prevent re uploading/downloading a file that hasn't changed but for some reasons has different modtimes. Where a backend has a modtime (which surely is the norm anyways?), conflict resolution should work by keeping the newest file unnamed to prevent any major breakage / damage to a running application that relies on these files. Also, this is a limitation of bidirectional sync its not designed to merge files so if somebody makes a change to file from both ends they should already expect a conflict. If this is something happening regularly in their workflow they need to change the way they're working e.g. do not change a file until it has updated to the latest copy or use git or switch over to using a central database or use libreoffice in the cloud, ms365 or google docs. I also believe this should be the default however if you believe strongly against this a flag would be greatly appreciated. My view on flags and why this should be the default is that advanced users should add flags to change the industry norms of a tool not the other way round. If using a remote that does not support modtime then sure maybe you have to keep both files named conflict1/path1 and conflict2/path2 as you can't determine which is newer.

## Conflict Naming

I see with the current way you would just end up with ..path1..path1 ..path1..path2 so it doesn't actually limit the number of conflicts (except that the file name could be come exceptionally long). 

> If the Path1 conflict were to have yet another conflict after that (and I'm not sure why that would happen, as it would mean opening x..path1 and editing it on both sides without renaming it first -- but let's just assume it did), you would then get x..path1..path1 and x..path1..path2, in addition to the pre-existing x..path2 which would remain unchanged.

Agreed, but just because you changed the file name back to the original doesn't mean you won't get another conflict, what if someone else has already handled the conflict.

> Does this question rely on the the prior suggestion (of renaming only one side) being implemented? 

Yes the suggestion to use conflict1, etc.. is with the fact that you would always keep the newest file. Also PATH1 and PATH2 is ultimately meaningless as by very nature if wanting to bisync a remote that means you are expecting more than one person/pc to update the remote and would also assume the use of the same tool e.g. rclone (in general). So really it would only have meaning if you also include a PC/PERSON unqiue ID too as PATH1 and PATH2 could be different in each case e.g. ..PATH1..COMPUTER1, ... Anyways, my point here is that PATH1 and PATH2 is meaningless in a bisync capacity so why use it at all? 

## Race Conditions

Yes, I was talking about a path because if there is a conflict we are now talking about multiple file(s) rather than a file. I do think using a lock on the file/path level would solve this. 

With regards to directories, I feel like what you are talking about here applies to all bidirectional file sync that aren't watching the directories and therefore do not have the concept of deleted_at to handle this case.

- Directory 'dir' is deleted on machine A
- On machine B somebody adds a file to 'dir' called 'file'

I'm not sure how rclone handles this? The two cases I can see is Machine A syncs first or Machine B

- Machine B adds 'dir.file' then Machine A delete 'dir' meaning 'dir.file' is no more
- Machine A delete 'dir' then Machine B add 'dir.file' 

Though, I thought rclone handles files rather than directories that's why there is a `--remove-empty-dir` flags which would mean even in the case listed above regardless of sync order 'dir.file' would survive and all files that were originally in 'dir' would be removed as one would expect? There not much a tool can do if lets say multiple people use bisync on a single drive where one deletes a folder and another one adds stuff to said folder. The best it can do is try to keep the new file and delete the old ones. Though, perhaps this is not the case with bisync and it would be entirely dependant on which machine syncs first. Either way, it up to users to not delete a directory which others are using otherwise they're gonna have to use there archives to restore data (fair enough no?) and why one has archives..

This also comes back to how one uses bisync. In the common case where someone is mass deleting directories, I believe it would be to keep multiple pc in sync for one user not on a shared drive used by others for this very reason. I know that if I want to delete 'dir' v1 and replace 'dir' with v2 I should : 

1. do on one machine 
2. if using another machine first wait for 'dir' to be deleted before recreating 'dir' with new files
3. do not make changes to 'dir' until up to date from step 1 or 2

> "Note that while concurrent bisync runs are allowed, be very cautious that there is no overlap in the trees being synched between concurrent runs, lest there be replicated files, deleted files and general mayhem."

This is a limitation that as far as I am aware of applies to all bidirectional syncing and everyone is aware so I still fail to understand if there is something else in 'bisync' that warrants this statement. Like the above is kinda of obvious no? Like yes, if you try and edit stuff you've deleted on another machine then get ready for some issues when it tries to sync up. Though, I still think the file you edited would remain as rclone uses files not directories. (perhaps only when --create-empty-directories is used would this create the issue where sync order would matter).


## Sessions (not sessions 🤣)

I agree, I'm using the wrong terminology here but ultimately have nothing more to add to your points. I'm in full agreeance and also yes believe there are more pressing matters at hand. I would also expect that when local this should add virtually no overhead. 


## Graceful Shutdown

My setup is as such to create separation of different crypt keys, storage remotes, usage and local file path/location. Also to avoid too many files in one place as bisync in its current state simply can't handle the load. So take for example: Cold, Documents, Work, VMs, this list goes on. Each one of these may reside on different remotes and my desktop for example keeps a full local copy of all where my laptop will mount drives that do not need to be local such as Cold. To make this work for me now I've already sacrificed storing files in certain directories so a sync for my Work directory for example can be done in 10 minutes as per the first comment of this issue rather than over an hour as it was taking before. One of my directories contains over 100,000 files and this could take 12 hours to sync. I would only do this manually until the performance is increased. I simply call the below script with all the systemd services when a shutdown is called which will wait until the service is inactive and then stop it completely. These service will notify if any of the syncs require a --resync or exit 1 errors via `notify-send` so I can assume sync is working otherwise.

```
isActiveRunning() {
	if systemctl --user is-active --quiet $1; then echo true;
	else echo false; fi
}

stopService() {
	while true;	do
		if [ "$(isActiveRunning $1)" == "false" ]; then
			systemctl --user stop $1
			echo "$1 stopped"
			break
		fi
		echo "$1 still running..."
		sleep 5
	done
}

stopService {serviceName}
``` 

Thanks !

**Comment by nielash** [comment]
> having a remote --workdir would require a complete change to --workdir to handle unique ids

You are right -- it would. This is work that would have to be done when implementing the feature.

The rest of your comments in this section describe why you personally do not wish to use a remote `--workdir`, which is fine -- you don't have to. There are others who do want such a feature. I think the only way it affects you is that it means we can't rely on PIDs as our locking mechanism -- but as I've explained, there would already be problems with doing so. At most, checking PIDs could offer some protection within the same machine, but it could not protect against the risk of race conditions caused by bisync runs on other machines.

> Also, a lock file is to do with the remote and not workdir.

I agree this is a valid point, and I think it may be worth considering a separation -- such that you can have a remote lock file and a local `--workdir`. I think that might be my ideal combination. 

> By very nature and for bisync to be fully functional requires a backend to support mod time. That's not to say that workarounds can't be put in place but simply when using bisync one should always be advised to use it with a backend that supports modtime.

I disagree -- please read the following:

- [5679 JWink3101 comment (author of bisync-alternative "syncrclone")](https://github.com/rclone/rclone/issues/5679#:~:text=The%20fact%20that,one%20side%20only.) 
- [How Bisync avoids needing modtimes to match precisely](https://rclone.org/bisync/#modification-time)
- [User feedback re Bisync's lack of backend support](https://forum.rclone.org/t/bisync-should-be-considered-experimental/29879)
- ["This limitation will be lifted in a future rclone bisync release."](https://rclone.org/bisync/#supported-backends)

I think where we do agree is that I personally would probably not ever use a backend that didn't support modtimes (and [there are many](https://rclone.org/overview/#features).) But that is just my personal preference. There are many other users who feel differently and do want to use such backends. There are even many users who go out of their way to [avoid using modtimes, even when the backend supports it](https://rclone.org/s3/#avoiding-head-requests-to-read-the-modification-time). This is popular with [S3-compatible backends](https://rclone.org/s3), for example, for performance and money reasons. Bisync should be available to them too.

Modtime is also imperfect -- it relies on the assumption that the underlying Fs is always updating it correctly, and that's not always true. It can be fooled (for example by using `touch`). It is not sensitive to bit rot. And even if it were perfect, "newer" does not always equal "better". There are times when a user does intentionally revert to an older version.

I think the best compromise here is a `--prefer newer` flag that renames conflicts in the way you describe if modtime is available for both paths.

> Also PATH1 and PATH2 is ultimately meaningless as by very nature if wanting to bisync a remote that means you are expecting more than one person/pc to update the remote and would also assume the use of the same tool e.g. rclone (in general). So really it would only have meaning if you also include a PC/PERSON unqiue ID too as PATH1 and PATH2 could be different in each case e.g. ..PATH1..COMPUTER1

This is a good point, and perhaps we should consider adding such a unique ID.

> - Directory 'dir' is deleted on machine A
> - On machine B somebody adds a file to 'dir' called 'file'
> I'm not sure how rclone handles this?

This is explained in [the docs](https://rclone.org/bisync/#unusual-sync-checks). In short: the non-deleted path always wins (when the other is new/changed). (But note that anything else in the dir, such as `dir/file2`, would still be deleted.)

> I guess I also struggle to see how concurrency changes the outcome of the sync if rclone is file based but perhaps I'm wrong about this.

When there's only one run at a time, it does not change the outcome of the sync. The problem is when there's more than one run at a time. The outcome then depends on which machine gets to that directory _first_, and _that_ is what cannot be predicted, because of concurrency (and potentially because of other things too, like differences in network speed, CPU/RAM, size of other jobs in the queue, etc). 

> Like the above is kinda of obvious no? Like yes, if you try and edit stuff you've deleted on another machine then get ready for some issues when it tries to sync up.

It's obvious to me, but I can't assume it will be obvious to everyone 😄 

But if it's obvious, then surely it is also obvious why a lock file must ideally exclude additional runs on _any_ machine, not just the _same_ machine.

> My setup is as such to create separation of different crypt keys, storage remotes, usage and local file path/location. Also to avoid too many files in one place as bisync in its current state simply can't handle the load.

So, it sounds to me like the reason you have so many simultaneous bisync runs is:

1. You have many different remotes
2. Bisync currently is slow

For 1, have you looked into the [Combine backend](https://rclone.org/combine/)? This is exactly what it's for.
For 2, I agree -- bisync is currently slow. I'm working on improving that.

Am I understanding correctly that if the speed issues were solved, and you combined your remotes with the Combine backend, you would no longer need multiple simultaneous bisync runs? Or is there an additional problem that I'm overlooking?

**Comment by crocinsocks** [comment]
## Workdir

I think the problem here is that there is only one use case for a remote --workdir and it requires no concept of unique ids or machines is syncing remote to remote (even if multiple remotes e.g. remoteA to remoteB, remoteC to remoteD, and so on). As soon as machine is syncing a local folder than there is 0 benefit to use a remote --workdir and instead only adds so much unneeded complexity. That is my only point and has nothing to do with PID as we've already spoken about this.

Take for example you configure workdir to be able to handle multiple PCs:

- I now want to use the remote workdir to sync remoteA and remoteB?
    - Using a unique ID only breaks this setup unless all remote to remote are stored globally as there is no relevance to which machine synced remoteA to remoteB.
- When would anyone ever be able to sync a local to remote of another PC and therefore need this information stored in a remote?
- I copy my VM and run a sync it will have the same ID so I now have a broken sync and likely corruption. Or same case that one duplicates the uniqueID?
- How do you correctly handle unique IDs, restorations alongside loss connectivity and failures?
- Why create more requests and complex workdir with no benefit?

There is only one "potential advantage" to allowing multiple PCs to use a --workdir for anything other than remote to remote and that is if my computer crashes due to hard drive failure and my data is stored on another drive which didn't crash I could recover the session cache provided I knew this unique ID. Though this a weak reason as this should be solved by backing up your cache directory for the event of failure (even if that is to a remote and using rclone)

All that would need to be done is to allow workdir to be remote when both paths are also remotes which should be very fast as no changes are required to workdir. I'm only saying this for your benefit as why spend hours coding what would be a useless feature with as far as I can tell would have an absolute zero gain.  

## Conflicts / ModTime

I can see the lack of mod time is useful when you already have a preferred direction to reduce cost and speed up transfers. Also agree if people are happy for bi sync with no preferred direction to just have files randomly renamed files on conflicts and no mechanism to determine latest fair play, I personally wouldn't but each to their own 🤣 my point just below may also play into this regarding symmetry as bisync progresses. 

A flag to prefer 'newer' is more than valid and may tie into the point I just added to resync in #5681 and perhaps this needs more thought in general about conflict resolution and bisync nature as either symmetrical or asymmetrical.

I also agree generally that modtime is not always perfect but in most cases it is. I also agree with the fact that a user may need access to a conflict which is why I'm very happy bisync already keeps conflicts by default. The newer flag in my case is not to prevent any other conflict file being made but to simply keep at least one of them always named the same.

The problem with then naming files with a unique ID is this is arguably only valid in a single user environment where they can still find useful info about the unique ID e.g. oh that's my laptop. But what in a multi user environment with many people syncing it still becomes rather unuseful. Which is why as far as I am aware most tools use .conflict1, .conflict2 naming. However, in a single user or small multi user environment I could see a conflict being named using some form of unqiueID could be useful.

## Race Conditions

> This is explained in [the docs](https://rclone.org/bisync/#unusual-sync-checks). In short: the non-deleted path always wins (when one path is deleted and the other is new/changed). (But note that anything else in the dir, such as dir/file2, would still be deleted.)

Exactly, what I though and what should happen. 

>When there's only one run at a time, it does not change the outcome of the sync. The problem is when there's more than one run at a time. The outcome then depends on which machine gets to that directory first, and that is what cannot be predicted, because of concurrency (and potentially because of other things too, like differences in network speed, CPU/RAM, size of other jobs in the queue, etc).

Sorry I am still confused as to why? If sync order has no effect then running concurrently should have no effect. Perhaps only if one is trying to delete an entire directory but if this is the only case then surely a check should be put in place to prevent the deletion of directory if files still remain even if the sync thinks it should delete that directory and has removed all the files.

> It's obvious to me, but I can't assume it will be obvious to everyone 😄
> But if it's obvious, then surely it is also obvious why a lock file must ideally exclude additional runs on any machine, not just the same machine. 

I believe the first one is obvious because that is how almost all bi sync tools work and people have experience with these from the most un techy all the way through to the super techs. Though of course to someone who has never used bisync tools perhaps not.

The reason the second one is not obvious is most bisync tools do run concurrently and the explanation is rather vague and does not list the cases that would cause said problems. As far as I can tell the only need for lock files is to prevent manipulating the same path/file at the same time. Though I am still trying to determine if there are any other reasons. The reason being is without actually knowing the race conditions one can't look at fixing these and I really dislike the idea of a global lock in my case as you know I'd rather a file lock. But would that even work. Without actually knowing the race condition of concurrent sync it would be impossible to answer. Perhaps, at this stage the answer to this is unknown and of course that's okay but I think for this tool to come out of experimental this must be known. 

## Multiple Processes

Yes, combined remotes could help but that does disregard different local directories e.g. I'm not storing all my cold data on SSD for example. For ones where they are all stored on same drive perhaps this could reduce the need of multiple processes once bisync can handle vast number of files e.g. millions. So thanks for the suggestion. At this stage, I have no issue with running multiple processes. At the end of the day and regardless of how many process as I would still have to call this script to ensure a graceful shutdown until such time as this is implemented into bisync or rclone more generally. 

Cheers !

**Comment by nielash** [comment]
Firstly, I agree with you about the technical challenges surrounding unique IDs. Fully agree that either these challenges must be solved, or the feature can't be implemented. You seem to assume they can't be solved -- I am not so quick to assume that. But if you turn out to be right that they are totally unsolvable, then of course the feature will not be implemented.

> As soon as machine is syncing a local folder than there is 0 benefit to use a remote --workdir and instead only adds so much unneeded complexity.

Even if I accept your premise here, you are still acknowledging that remote-to-remote is a valid use case. So, if the feature is being built anyway for remote-to-remote, then the question becomes: should we disable the already developed feature for local? Or allow each user to make their own choice, in case they have some use case we haven't thought of? I lean towards the latter, even though I agree that most users (including me) would probably choose not to use it.

One possible use I can think of is if you want your listing files to be encrypted at rest, since they contain a list of all of your files in plaintext, even files from a `crypt` remote. If remote `--workdir` is supported, you could store these listing files on a `crypt` remote. Otherwise, you have no choice but to store them in plaintext on your local machine. 

The bottom line is: I don't claim to know every possible use case that a user might have, and so I think it's better to make the option available, even if it might not get used much or at all (assuming the dev work is mostly happening anyway).

> Also agree if people are happy for bi sync with no preferred direction to just have files randomly renamed files on conflicts and no mechanism to determine latest fair play, I personally wouldn't but each to their own 🤣 

I don't agree that it's "random" or that there is "no mechanism to determine latest". The beauty of the ["Dual Shadow Method"](https://rclone.org/bisync/#references) is that you actually _don't_ need modtimes to determine "latest", usually. For example, if you only have checksums, if the checksum changed on Path1 but stayed unchanged on Path2, you can conclude that Path1 is "latest" and should be synced to Path2.

> As far as I can tell the only need for lock files is to prevent manipulating the same path/file at the same time. Though I am still trying to determine if there are any other reasons. The reason being is without actually knowing the race conditions one can't look at fixing these and I really dislike the idea of a global lock in my case as you know I'd rather a file lock. But would that even work. Without actually knowing the race condition of concurrent sync it would be impossible to answer. Perhaps, at this stage the answer to this is unknown and of course that's okay but I think for this tool to come out of experimental this must be known.

The race condition scenarios are very much known, and I've been trying to explain them. Since you have acknowledged that you understand the problem with manipulating the same path/file at the same time, I will skip that part of the explanation and just take that as a given. So, the next question is how concurrency could cause that to happen. Well, imagine that you have two machines syncing at overlapping times, but the directory order and completion time are perfectly predictable. In this situation, there might not be a problem, because while the sync as a whole overlaps, no two directories actually overlap. We can conclude that Machine 1 will always sync any given directory before Machine 2:

```
Machine 1: | A | B | C | D | E | F | G | H | I | J | K | L |
Machine 2:         | A | B | C | D | E | F | G | H | I | J | K | L |
```

Now, imagine the same thing with an unpredictable sync order. Now, it becomes possible that Machine 2 will sync a given directory at the same time or earlier than Machine 1, as noted by the `!` symbol:

```
Machine 1: | I | J | E | D | F | C | L | A | K | H | B | G |
Machine 2:         | D | L | G | J | A | H | I | C | B | E | K | F |
                     !   !   !       !   !           !
```

But with a mutual exclusion lock, the random directory order isn't a problem, since it can never overlap:
```
Machine 1: | I | J | E | D | F | C | L | A | K | H | B | G |
Machine 2:                                                 | D | L | G | J | A | H | I | C | B | E | K | F |
```

Hopefully that clarifies the problem.

Additionally, [here is a simple little example](https://go.dev/play/p/S98GjeaGBX0) illustrating how order can be unpredictable due to concurrency in Go. Click "Run" multiple times, and observe how the worker numbers are printed in a different order each time, even though the code has not changed.

> Yes, combined remotes could help but that does disregard different local directories e.g. I'm not storing all my cold data on SSD for example.

I'm not sure if I'm understanding what the type of drive has to do with this. The [combine backend](https://rclone.org/combine/) is perfectly capable of combining SSD and HDD paths into one virtual tree for rclone's purposes. (They would still remain on their original drives -- rclone would just see them as part of the same tree. Similar to symlinks.) But in any case, I am glad you found a solution that works for you.

**Comment by nielash** [comment]
> Just to clarify I'm not saying they can't be solved I'm saying that there is no advantage and a lot of complication in doing so

I hear you, and I don't have much more to add on this topic. I think we agree on the "a lot of complication" part, disagree on the "no advantage" part. 

> Remote-to-remote requires no changes to workdir though?

It would still require changes, as currently all of this logic uses the `os` module. It actually would be _harder_ to not change local too, because it would require having two totally separate code paths for remote vs. local (vs. just one for remote, which could just happen to be a remote of type `local`.) So the question is really: should we _go out of our way to disable_ access to a feature that's already built? (Although you are right that it's not quite that simple, since enabling it also adds some complexity.) 

> Other than remote to remote (which is easy and I agree should be added) the files are already present on the machine so anyone who is already on your machine will have this data anyways.

It could be `local` wrapped by `crypt`, meaning the local files are encrypted and only readable by rclone. No, this will never be resistant to a "$5 wrench attack", but I still think it's better than nothing -- in addition to obtaining your files, someone would have to obtain your crypt password / config and understand how to use rclone to decrypt it. Not impossible, but it would take a lot more effort. 

> Sorry if it wasn't clear I was talking about conflicts only and that in bidirectional sync you would just have randomly renamed conflict files with no chance of a winner. Which I was saying for me is enough of a reason to never use a non modtime backend for bisync.

But that's what you already have now with bisync, whether your backend supports modtime or not. As I mentioned earlier, I see it as a feature--not a bug--that bisync refuses to engage in "guessing" about winners and losers.

> I think I am understanding you and that I am right then in saying that the only issue with concurrency is writing to the same path/file at the same time and that is it?

In general that is correct, but there are some important caveats to understand:

- Currently bisync builds its entire list at the beginning, and does all its "checking" before transfers start. It then filters the transfers for only those files that it determines need to be synced. Those files get re-checked the moment before they are transferred, while the other files never get re-checked. There can therefore sometimes be a fairly large window between when bisync gets information and acts on that information, and thus more chance of bisync finding something unexpected by the time it gets there (ex. trying to delete a file that's already gone)
- As we've discussed, bisync currently saves the state only once at the very end -- not after each transfer.
- If you use `--check-sync=true` (which is the default), the entire sync will fail and require `--resync` if the two sides don't match exactly at the end of the sync. This could easily be triggered by a change from another machine. (Even your Example 4 where the trees overlap but the changed paths do not.)
- "The same time" could be quite a long window, in some scenarios (ex. huge file that takes hours to transfer)
- How you define "path" here matters (see below)

> As sync order doesn't matter as long as its not the same path at precisely the same time everything should be fine e.g. so a lockfile on the file/path level would suffice?

Well, how do you define "path"? Is a "path" still a "path" if it has "sub-paths"? Isn't your root directory still a "path"? This is where it starts to make a big difference whether atomicity is important to any of your directories. If we care only about individual files, your chances of a collision are much smaller (but still not zero). On the other hand, if we care about directory-level atomicity, then basically any amount of overlap is a potential hazard. After all, who's to say at what specific hierarchy-level the atomicity should start to matter? Isn't it possible that some users bisync a root path that they consider to be atomic? All to say: a file-level lock file would work if you only care about files with no concern for directory atomicity. If you do care at all about directory atomicity, the only viable lock file is one at the level of your bisync root path. 

> Though following on you do say Machine 2 may get there earlier and I fail to see a problem with this as sync order does not matter.

Sync order does matter. Your Example 3 is a good example (file deleted and changed at same time). If Machine A gets there first, the file is deleted by A and then re-added by B, causing it to propagate back to A on the next run. But if B gets there first, the file is updated by B and then deleted by A, causing the delete to propagate back to B on the next run. One way the file survives everywhere, the other way it gets deleted everywhere. Only difference is which machine gets there first. 

I think that is the best that I can explain it... if it still doesn't make sense, perhaps it would be helpful to test it (not with any files you care about 😅 )
